pr_agent/git_providers/local_git_provider.py

```diff
@@ -0,0 +1,201 @@
+import logging
+import uuid
+from collections import Counter
+from pathlib import Path
+from typing import List
+
+from git import GitCommandError, Repo
+
+from pr_agent.config_loader import _find_repository_root, settings
+from pr_agent.git_providers.git_provider import EDIT_TYPE, FilePatchInfo, GitProvider
+
+
+class PullRequestMimic:
+    """
+    This class mimics the PullRequest class from the PyGithub library for the LocalGitProvider.
+    """
+
+    def __init__(self, title: str, diff_files: List[FilePatchInfo]):
+        self.title = title
+        self.diff_files = diff_files
+
+
+class LocalGitProvider(GitProvider):
+    """
+    This class implements the GitProvider interface for local git repositories.
+    It mimics the PR functionality of the GitProvider interface,
+    but does not require a hosted git repository.
+    Instead of providing a PR url, the user provides a local branch path to generate a diff-patch.
+    For the MVP it only supports the /review and /describe capabilities.
+    """
+
+    def __init__(self, branch_name, incremental=False):
+        self.repo_path = _find_repository_root()
+        self.repo = Repo(self.repo_path)
+        self.head_branch_name = self.repo.head.ref.name
+        self.branch_name = branch_name
+        self.tmp_branch_name = f'pr_agent_{uuid.uuid4()}'
```

Comment:
The diff between HEAD and the target_branch will not be equal to the "PR diff, changed files", unless we run `git rebase target_branch` first.
Otherwise, in case the target_branch was developed since we branched out, our local diff will contain changes stemming from the target_branch development (PR changes wouldn't).
We need to run `git rebase target_branch` on our HEAD, in order for `git diff target_branch` to be equivalent to the PR diff.
We create the tmp_branch to perform rebase there, and not impact the original HEAD in any way

Reply:
<img width="751" alt="image" src="https://github.com/Codium-ai/pr-agent/assets/4869233/fd11dc04-1307-4503-9128-d227039d4e8a">

Usually when preparing a PR I'll want to review my changes - the difference from the merge base and my head brange (three dot diff), and this is the default behavior in GitHub, that's why I ask. When we merge, if the target branch was updated, which is also not typical in my local copy, I'd want to rebase my head branch first and then start the review, and deal with the merge locally. That's why I think a diff to the merge-base is better.

Reply:
In other words, I suggest to take the diff between 'head' and  repo.merge_base(head, target)

Reply:
I agree, three-dot git diff makes more sense, and will reduce much of the complexity. I will change that, thanks.